### PR TITLE
[WIP][MIG][11] Rough migration of asterisk_click2dial, base_phone, base_phone_popup, crm_phone and event_phone

### PR DESCRIPTION
**Pull request by sadasant**
Hi! I'm making this PR just in case anyone finds it useful. I need to have this working ASAP so I'm making general changes without much detail. Once it works for us, I'll take the time to follow up the community guides and I'll make a proper PR.

Please feel free to close this PR and leave any form of comment, I'll respond and I'll listen to any instruction additional to https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-11.0
 
## Changes:

Here's a list of the changes I've made. It's here to help highlight the commits, and give more information about them. Lines containing **NOT SURE** are about things that **I'm not so sure if they are correct**, but they seem to work.

1. Commands I initially used:

```
git checkout -b 11.0-mig-some-modules origin/11.0
cp -fr ../connector-telephony-10/asterisk_click2dial ./
cp -fr ../connector-telephony-10/base_phone ./
cp -fr ../connector-telephony-10/base_phone_popup ./
cp -fr ../connector-telephony-10/crm_phone ./
cp -fr ../connector-telephony-10/event_phone ./
cd asterisk_click2dial/
2to3- -wnj4 --no-diffs .
cd ..
cd base_phone/
2to3- -wnj4 --no-diffs .
cd ../
cd base_phone_popup/
2to3- -wnj4 --no-diffs .
cd ..
cd crm_phone/
2to3- -wnj4 --no-diffs .
cd ../
cd event_phone/
2to3 -wnj4 --no-diffs .
2to3- -wnj4 --no-diffs .
cd ..
cp -R ../connector-telephony-10/setup/asterisk_click2dial setup/
cp -R ../connector-telephony-10/setup/base_phone setup/
cp -R ../connector-telephony-10/setup/crm_phone setup/
cp -R ../connector-telephony-10/setup/event_phone setup/
find . -name '*.po' -print0 | xargs -0 sed -i "" "s/ir\.actions\.report\.xml/ir.actions.report/g"
```

2. Then I removed these lines:

https://github.com/OCA/connector-telephony/blob/10.0/base_phone/i18n/fi.po#L763-L791
 
3. [887b356](https://github.com/OCA/connector-telephony/pull/156/commits/887b35674c79aa13f9a24c29488a12a3294f8e93) Replaced `base.config.settings` with `res.config.settings` in all these files:

```
base_phone/i18n/de.po
base_phone/i18n/es.po
base_phone/i18n/fr.po
base_phone/i18n/sl.po
base_phone/i18n/tr.po
base_phone/models/base_config_settings.py
base_phone/views/base_config_settings.xml
```

4. [d0cde6f](https://github.com/OCA/connector-telephony/pull/156/commits/d0cde6f8ca63b29fe81ac4e81bd57a81394f041f) replaced `base_setup.view_general_configuration` with `base_setup.res_config_settings_view_form` in `base_phone/views/base_config_settings.xml` **NOT SURE**.

5. [0efe5e8](https://github.com/OCA/connector-telephony/pull/156/commits/0efe5e81f1994923f8479fcb3c2428063e8f1532) Replaced `<group name="report" position="before">` with `<xpath expr="." position="inside">` in `base_phone/views/base_config_settings.xml` **NOT SURE**, the `expr="."` looks dangerous 😂.

6. [fde24e6](https://github.com/OCA/connector-telephony/pull/156/commits/fde24e64b9b37786c73b56d660a90f60e36fbd2a) Replaced `sales_team.menu_base_partner` with `sales_team.menu_sale_config` in `crm_phone/view/crm_phonecall.xml`, **NOT SURE**.

7. [6d6a010](https://github.com/OCA/connector-telephony/pull/156/commits/6d6a01095389416de399408ff12845e83f7faefa) Replaced `reformat.all.phonenumbers` with `phone.reformatphonenumbers` since `reformat` doesn't exist anymore. It works, but might not be the best approach.

8. [6f70123](https://github.com/OCA/connector-telephony/pull/156/commits/6f701230783671753ae71f6507f8d140ff4486ac) Replaced `reformat.all.phonenumbers` with `phone.reformatphonenumbers` in all the matching `.po` files.

9. [03aaec7](https://github.com/OCA/connector-telephony/pull/156/commits/03aaec73319646b8ba847951e687cb7eb0b7db13) Added `sale.` to the parent reference in `menuitem id="crm_phonecall_menu" name="Phone Calls` in `crm_phone/view/crm_phonecall.xml`.

10. [6fb0f18](https://github.com/OCA/connector-telephony/pull/156/commits/6fb0f18fbf0b3f4858b69c6e2ed404400ca08b32) This commit includes several changes, some of them were overwritten by later commits, but the highlights are:

- Replaced `CrmPhonecall` with `PhoneCall` in `crm_phone/models/crm_phonecall.py`.
- Replaced `sales_team.menu_sale_config` with `sale.sale_menu_root` in `crm_phone/view/crm_phonecall.xml`, this first started in [this commit](https://github.com/OCA/connector-telephony/pull/156/commits/6fb0f18fbf0b3f4858b69c6e2ed404400ca08b32), but later it became clear that I needed to add `sale.` at the beginning.
- Replaced `WizardCreateCrmPhonecall` with `WizardCreatePhoneCall` for consistency, but perhaps it has a real effect 👀.

11. [bd17aab](https://github.com/OCA/connector-telephony/pull/156/commits/bd17aabc64eb7664be2581cc823b60926b6c6a2f) this is a merge commit with several changes, the highlight is that I replaced `core.list_widget_registry` with `core.search_widget_registry` in `base_phone/static/src/js/phone_widget.js`.

12. [7c9a0f3](https://github.com/OCA/connector-telephony/pull/156/commits/7c9a0f3c287f1931387277fba1a97550f52a26df) This commit makes `base_phone` and `asterisk_click2dial` generally work, it includes the following changes:

- Replace all the RPC calls in the `static/src/js` folders to use `ajax.jsonRpc` from `web.ajax`.
- Replace the uses of `phonenumbers.parse` to the following code, that attempts to parse it with no region, but if it fails tries again with `region='US'`:

```python
try:
    res_parse = phonenumbers.parse(res, None)
except:
    res_parse = phonenumbers.parse(res, region='US')
```

- Also started to catch exceptions in `convert_to_cache` in `base_phone/fields.py`.
- Fixed `.translate(None, ' -.()/')` to work in Python3 with `.translate(str.maketrans('','', ' -.()/'))`.
- Added constant functions `x => x` as the values for `field_utils.format.phone` and `field_utils.parse.phone = x => x` as a little hack to make both formatting and validation possible. Otherwise we would get huge errors on the rendering, editing and saving these fields. This was painful to debug 😂.
- Added `formatType: 'phone'` to FieldFax, relevant to field_utils' `format` and `parse` objects.
- Removed `var ColumnPhone = Widget.extend({` just to reduce the number of changes to have something working.
- Odoo 11 has a `FieldPhone` which we're entirely replacing with `field_registry.add('phone', FieldPhone)` in `base_phone/static/src/js/phone_widget.js`.
- FieldFax is not using the template anymore, now we're relying on `className: 'o_field_phone'`
- Removed `initialize_content`.
- Moved the functions assigned to `render_value` to `_renderReadonly` (there's an [extra return](https://github.com/OCA/connector-telephony/pull/156/commits/7c9a0f3c287f1931387277fba1a97550f52a26df#diff-021c722bd9e56d70b5153bac2115edc2R51) in FieldFax's _renderReadonly there that I removed in [f010bac](https://github.com/OCA/connector-telephony/pull/156/commits/f010bace369956bdff2f2a985faecbccd3bc4a25)).
- `FieldPhone` has a new `isValid` function that tests the value against a simple [RegExp](https://github.com/OCA/connector-telephony/pull/156/commits/7c9a0f3c287f1931387277fba1a97550f52a26df#diff-021c722bd9e56d70b5153bac2115edc2R65) for validating international phone numbers. The RegExp is not 100% accurate, as one can expect.
- FieldPhone's _renderReadonly no longer uses `this.$el.filter('#click2dial')` but `this.$el` instead, since it was far easier than trying to make sure `#click2dial` was being rendered 🙈.
- FieldPhone's _renderReadonly now does `this.$el.removeClass('o_text_overflow')` just as some other fields do in Odoo 11.
- FieldPhone's _renderReadonly's click event now does `ev.preventDefault();`.
- FieldPhone's _renderReadonly now gets the model from `self.model` instead of ` self.view.dataset.model`.
- FieldPhone's _renderReadonly now gets the ID from `self.recordData.id` instead of `self.view.datarecord.id`.
- Also replaced all the RPC calls in FieldPhone's _renderReadonly to use `ajax.jsonRpc` from `web.ajax`.
- Made it so we're forcefully adding FieldFax and FieldPhone to `field_registry` since there's a FieldPhone already in Odoo.
- Removed the contents of `base_phone/static/src/xml/phone.xml` 🙈.

13. Fixed _getValue to retrieve the $el.text if there is no $input.

### CODE REVIEW

**Comment by sadasant** [comment]
I'll close this myself until I have something better to show.

**Comment by sadasant** [comment]
This seems to be working :) I know it's rough but let me know if this can be helpful in any way. I'm available for polishing the details, but feel free to override or use these changes at will!

**Comment by cmsalmeida** [comment]
Hi
Module crm_phone raise an error if sale module is not installed, installing sale module before, solves the issue.
Maybe you should add sale in depends, or fix crm_phone/view/crm_phonecall.xml:128
 - replace sale.sale_menu_root by crm.crm_menu_root

I've tested and works!

PS:

Plase check https://github.com/sadasant/connector-telephony/pull/1

**Comment by jchanme** [comment]
Is this PR working in Odoo 11 now?

File: base_phone/wizard/reformat_all_phonenumbers_view.xml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+    <!-- Menu entry under Settings > Technical -->
+    <menuitem id="menu_config_phone" name="Telephony" parent="base.menu_custom"/>
+
+    <menuitem id="reformat_all_phonenumbers_menu" action="reformat_all_phonenumbers_action" parent="menu_config_phone" sequence="100"/>
+
+    <!-- Open the Reformat Phone Numbers wizard after the installation of the module -->
+    <record id="reformat_all_phonenumbers_module_install" model="ir.actions.todo">
+        <field name="action_id" ref="reformat_all_phonenumbers_action"/>
+        <field name="type">automatic</field>
```
Comment (naglis) [inline base_phone/wizard/reformat_all_phonenumbers_view.xml:49]: Got a warning while installing:
```
odoo.models: ir.actions.todo.create() includes unknown fields: type
```
Seems `ir.actions.todo` [no longer has](https://github.com/odoo/odoo/blob/ced9156a97ea502050b4e827ef2237fa2cd9410f/odoo/addons/base/ir/ir_actions.py#L616-L619) a `type` field.

**Comment by pedrobaeza** [comment]
Closing as this:

* Doesn't preserve commit history.
* Mixes several modules in one PR.
* CIs are red.

If you want to contribute with the migration, follow the migration guide (https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-11.0), check CIs and make only one module per PR. Anyhow, we are now migrating `base_phone_popup`.

**Pull request closed without merging**